### PR TITLE
Feature/calendar month name

### PR DESCRIPTION
**Pull request by JakubSerafin**
Calendar view was missing name of the month and the year that user is actually browsing, so I've added it. .

Test for that functionality added.

Bit of stilling (baseline, arrows) improved in UI.

### CODE REVIEW

File: calendar/page.js [hunk 1]
```diff
@@ -26,12 +26,17 @@ function isRecordValid(record) {
   typeof record.title === 'string'
 }
 
+function getMonthName(){
+  return calendarHandler.calendar.getDate().d.d.toLocaleString('en-us', { month: 'long', year:'numeric' })
```
Comment (paulfitz) [inline calendar/page.js:30]: what is .d.d?

Comment (paulfitz) [reply calendar/page.js:30]: most grist code leaves a space after a colon

Comment (paulfitz) [reply calendar/page.js:30]: and a space before a curly brace starting a function (like `getMonthName() {`)

File: calendar/page.js [hunk 2]
```diff
@@ -82,31 +88,38 @@ class CalendarHandler {
       this.calendar.updateEvent(record.id, CALENDAR_NAME, {backgroundColor: CalendarHandler._selectedColor});
       this._selectedRecordId = record.id;
       this.calendar.setDate(record.startDate);
-      const dom = document.querySelector('.toastui-calendar-time');
-      const middleHour = record.startDate.getHours()
-        + (record.endDate.getHours() - record.startDate.getHours()) / 2;
-      dom.scrollTo({top: (dom.clientHeight / 24) * middleHour, behavior: 'smooth'});
+      if (this.calendar.getViewName() !== 'month') {
+        //Scoll to the middle of the event, if it's not month view
```
Comment (paulfitz) [inline calendar/page.js:92]: "//Scoll" -> "// Scroll"

What is this trying to solve? (add answer as a comment in code please)

File: calendar/page.js [hunk 3]
```diff
@@ -82,31 +88,38 @@ class CalendarHandler {
...
   // change calendar perspective between week, month and day.
   changeView(calendarViewPerspective) {
     this.calendar.changeView(calendarViewPerspective);
+    document.getElementById('calendar-title').innerText = getMonthName();
   }
 
   // navigate to the previous time period
   calendarPrevious() {
     this.calendar.prev();
+    setMonthNameInDom();
```
Comment (paulfitz) [inline calendar/page.js:110]: Perhaps a more general `updateAfterNavigation` name would work, could imagine needing to adapt the UI in several ways.

File: calendar/page.js [hunk 4]
```diff
@@ -82,31 +88,38 @@ class CalendarHandler {
...
   calendarNext() {
     this.calendar.next();
+    setMonthNameInDom();
   }
 
   //navigate to today
   calendarToday() {
     this.calendar.today();
+    setMonthNameInDom();
   }
```
Comment (paulfitz) [inline calendar/page.js:124]: Aren't there going to be other ways the period could change? E.g. via linking? Trying out linking locally, I can jump between months without the month name being updated.

Comment (JakubSerafin) [reply calendar/page.js:124]: Yes, it was a case that I've missed

File: calendar/page.js (hunk 1 above)
Comment (JakubSerafin) [reply calendar/page.js]: > what is .d.d?

TUI Calendar is using custom TZDate structure to handle the dateTime and  i used [..].d.d.[...] to access plain old Date format. Nerveless i found more canonical way to do this is documentation :)

**Comment by JakubSerafin** [comment]
> Test failure is replicable with MOCHA_WEBDRIVER_HEADLESS=1

even with a flag on i was not able to reproduce error on my machine, despite running tests multiple time. It's disturbing

File: calendar/page.js [hunk 5]
```diff
@@ -88,8 +88,11 @@ class CalendarHandler {
       this.calendar.updateEvent(record.id, CALENDAR_NAME, {backgroundColor: CalendarHandler._selectedColor});
       this._selectedRecordId = record.id;
       this.calendar.setDate(record.startDate);
+      updateUIAfterNavigation();
       if (this.calendar.getViewName() !== 'month') {
-        //Scoll to the middle of the event, if it's not month view
+        //Scroll to the middle of the event if it's not month view.
```
Comment (paulfitz) [inline calendar/page.js:93]: Thanks! Could you please add a space after the `//`s, to match the bulk of Grist source code.

File: calendar/page.js [hunk 6]
```diff
@@ -194,7 +197,8 @@ function getGristOptions() {
 }
 
 
-function setMonthNameInDom(){
+function updateUIAfterNavigation(){
+  //update name of the month and year displaed on the top of the widget
```
Comment (paulfitz) [inline calendar/page.js:201]: displaed -> displayed
[also spaces after // would save me time deciding whether it is a big enough deal to mention :)]

File: calendar/page.js [hunk 7]
```diff
@@ -82,31 +88,41 @@ class CalendarHandler {
...
+          + (record.endDate.getHours() - record.startDate.getHours()) / 2;
+        dom.scrollTo({top: (dom.clientHeight / 24) * middleHour, behavior: 'smooth'});
+      }
     }
   }
 
   // change calendar perspective between week, month and day.
   changeView(calendarViewPerspective) {
     this.calendar.changeView(calendarViewPerspective);
+    document.getElementById('calendar-title').innerText = getMonthName();
```
Comment (paulfitz) [inline calendar/page.js:107]: Could this be a call to `updateUIAfterNavigation`

**Review by paulfitz — COMMENTED** [review]
A few very small comments, pretty much there.

I can't replicate test failure on latest version, whereas 2acd12c failed consistently for me in headless.

File: calendar/page.js (hunk 7 above)
Comment (JakubSerafin) [reply calendar/page.js:107]: In current moment there is no point - changing perspective should persist day that calendar is focused on, so it should not change neither month nor year that is showed in header. But in future other UI element can be updated after navigation, so I'll add it

Comment (paulfitz) [reply calendar/page.js:107]: I'm saying that if you think of `updateUIAfterNavigation` as being in charge of updating the UI after some significant change, there's no need to duplicate the `document.getElementById('calendar-title').innerText = getMonthName();` code here - that can be the responsibility of `updateUIAfterNavigation`.

**Comment by paulfitz** [comment]
There's now a "changes you made may not be saved" style alert when trying to switch to another test.

Screenshot from 2023-09-12 09-17-21

**Comment by paulfitz** [comment]
@JakubSerafin I looked at what grist-core does and moved some code (https://github.com/gristlabs/grist-core/pull/677).

I see that this kind of alert is dismissed during simulated logins when working with an external server, which is like our situation in grist-widget:
https://github.com/gristlabs/grist-core/blob/bc6f5504712e47c19ab6b3e9244affc508d63812/test/nbrowser/homeUtil.ts#L115
I'd be interested in knowing what this is about, but it doesn't seem specifically a grist-widget issue - and commenting out calendar tests entirely, I can still see failures due to it.

So I just added a handler, like you did but at a wider scope.

**Comment by paulfitz** [comment]
@JakubSerafin btw I noticed when editing it that spacing is off in test/calendar.ts (4 versus 2)

**Pull request merged**